tests/SpecTests/DocumentsMatchConstraint.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+namespace MongoDB\Tests\SpecTests;
+
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use PHPUnit\Framework\Constraint\Constraint;
+use ArrayObject;
+use InvalidArgumentException;
+use RuntimeException;
+use stdClass;
+
+/**
+ * Constraint that checks if one document matches another.
+ *
+ * The expected value is passed in the constructor.
+ */
+class DocumentsMatchConstraint extends Constraint
+{
+    private $ignoreExtraKeys = false;
+    /* TODO: This is not currently used, but was preserved from the design of
+     * TestCase::assertMatchesDocument(), which would sort keys and then compare
+     * documents as JSON strings. If the TODO item in matches() is implemented
+     * to make document comparisons more efficient, we may consider supporting
+     * this option. */
+    private $sortKeys = false;
+    private $value;
+
+    /**
+     * Creates a new constraint.
+     *
+     * @param array|object $value
+     * @param boolean      $ignoreExtraKeys If true, ignore extra keys within actual value's document(s)
```

Comment:
This may need to be split into separate flags for ignoring top-level keys alone and extra keys throughout all sub-documents. Per my comment in https://github.com/mongodb/specifications/pull/560#pullrequestreview-246810295:

> In the Command Monitoring spec test README, [Additional Values](https://github.com/mongodb/specifications/blob/master/source/command-monitoring/tests/README.rst#additional-values) explicitly states that the allowance for extra fields only pertains to top-level fields, and an example is provided for a failed comparison due to an extra field within a sub-document.
>
> This is in contrast to the [Spec Test Match Function](https://github.com/mongodb/specifications/blob/master/source/change-streams/tests/README.rst#spec-test-match-function) described in the Change Streams spec test README, which I believe some drivers use as a general matching algorithm for documents in other spec's tests.

Reply:
Done in subsequent commit.